### PR TITLE
Fixing how stdDev in calibration is calculated

### LINKED ISSUE #19803: Standard Deviation calculation of calibrateCamera incorrect
##### System information (version)

- OpenCV => 4.5.1
- Operating System / Platform => Ubuntu 64 Bit
- Compiler => gcc

##### Detailed description

The standard deviation calculation of calibrateCamera is returning NaN for all standard deviations in all cases when provided with comparably low numbers of points.
This is due to [this line](https://github.com/opencv/opencv/blob/3e1673e8b2d5c15b028fb81f7413f5321daa8a2e/modules/calib3d/src/calibration.cpp#L1808), where the divisor and thus `sigma2` can become negative. The variable `total` usually is `nimages*maxPoints` or less, `nparams = nimages*6+18` (for standard calibration method), and `nparams_nz < nparams`.
For the case `total < nparams_nz < nparams` the resulting `sigma2` is negative.
This can (and nearly always does) happen when `maxPoints = 6` so `total/nparams < 1.`
This is likely also why it has been unnoticed so far.

Changing the `total` to `nparams` yields seemingly correct results. **EDIT:** Only for my case where `maxPoints = 6`! This might not be a general or even correct fix!

I've also opened a [forum question](https://forum.opencv.org/t/potentially-wrong-calibratecamera-standard-deviation-calculation/2461) previously, although I'm positive it is a bug.

##### Steps to reproduce

1. Generate a sample set of 22 6-point markers
2. Call `calibrateCameraRO`, requesting the standard deviations to be calculated
  - One sample run yielded the following variable values: `total=132 (22*6), nparams=150(22*6+18), nparams_nz=139`
3. Observe the returned standard deviations to all be NaN

An example implementation can be found in this larger project ([calibrateCameraRO call here](https://github.com/Seneral/MakersVR/blob/master/MakersVR_Configurator/src/calibration.cpp#L691)).

##### Issue submission checklist

 - [ X] I report the issue, it's not a question
   
 - [X ] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   
 - [ X] I updated to latest OpenCV version and the issue is still there
   
 - [ X] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by savuor**
Fixes #19803 

This PR should fix the standard deviation calculation for calibration procedure.

### TODOs:
- [x] Agree on the root cause of the problem
- [x] (Needs discussion) `total` should contain 2x the number of object points (one per each coordinate)
- [x] Forbid calibration calculation when a number of variables is less than a total number of residuals
- [x] Forbid stddev calculation when a number of variables is greater than or equal to a total number of residuals
- [x] Fix backport to 4.x
- [x] Make backport for 3.x

**EDIT:** the previous stddev calculation motivation is considered to be wrong, we need that denominator correction to make our estimates unbiased. An explanation of it can be found in [MVG].
Also, Cramer-Rao bound has nothing to do with it.

<details><summary>Previous incorrect explanation</summary>

_The stddevs estimation is based on [Cramer-Rao bound](https://en.wikipedia.org/wiki/Cram%C3%A9r%E2%80%93Rao_bound). While it's not said about the variance of samples,  the original code (Camera Calibration Toolbox by Jean-Yves Bouguet) and other available implementations calculate sigma2 based on all object points regardless how many are of them or how many parameters are to optimize._

</details>

I invite anyone to discuss that and verify that, especially those who've participated in the original issue discussion:
@Seneral @sanketc001 @lamm45 @alalek

I also invite @vpisarev and @ivashmak because of their experience in calibration.

* **MVG:** _R. Hartley, A. Zisserman, Multiple View Geometry in Computer Vision, 2004, section 5.1.3, page 134_

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
Why not 3.4 or 4.x?

**Comment by savuor** [comment]
@asmorkalov backport to 4.x: #22996

**Comment by lamm45** [comment]
Thank you for the invitation! I still think that the existing formula is correct. The proposed formula for sigma2 is only asymptotically correct (that is, when the number of points is huge).

As a "proof", the existing formula seems to be equivalent to the equation (1) of this paper: https://arxiv.org/abs/2107.13484

The issue isn't really specific to camera calibration. Here is a general explanation of why the number of parameters needs to be included in the denominator: https://stats.stackexchange.com/a/64495

I find the Cramer-Rao lower bound a bit confusing and unrelated to this discussion.

**Comment by savuor** [comment]
@lamm45 Thanks for your contribution!

Not sure if I've really got that explanation on StackExchange (while I've found another reference to it in the Multiple View Geometry in Computer Vision book, see section 5.1.3 at page 134).

Anyway, if we believe in it, then the best solution would be to forbid the calibration when `total < nparams_nz` since all the calibration makes no sense.
And to leave stddevs empty when `total == nparams_nz`.

What do you think?

**Comment by vpisarev** [comment]
my memories about this part of statistics are very fuzzy, but if I recall correctly the formula you propose is sort of biased estimate of the standard deviation. I would keep the original formula and, as you suggest, avoided cases when total is less than nparams_nz.

**Comment by lamm45** [comment]
> (Maybe this condition is incorrect since each point gives at least 2 equations - one per coordinate - and the prohibition bound should be twice bigger here)

I think you are right, the error estimator should use `2*total`. And not just for the prohibition bound, but for the actual error computation as well. So basically all estimates have been wrong by up to a factor of two, which is quite significant.

**Comment by savuor** [comment]
I've finally understood that denominator correction. While it's more related to ordinary least squares estimation variance, I also found a more general explanation of it in MVG book (see the reference in the PR description).
This explanation also contain a note that is strictly related to the calibration problem: when fitting model in the projection of N points, a dimensionality of error should be 2N, not N.
At the same time, I think that error computation is OK. It's calculation is based on a vector size, not on a `total` variable.
So basically the only place where `total` should be fixed is stddev estimation.

**Comment by lamm45** [comment]
I should have been more precise, but when I said error estimator or computation, I meant the estimator or computation for the *standard error*. In this context, standard error is essentially the same as stddev, but I prefer the former.

**Comment by savuor** [comment]
@asmorkalov Backport to 3.4: #23025

**Review by vpisarev — APPROVED** [review]

**Review by alalek — CHANGES_REQUESTED** [review]
Need align 3.4 / 4.x and this 5.x PRs to minimize conflicts.

**Comment by savuor** [comment]
@alalek Backport to 4.x is aligned with 5.x code, please check: #22996
